### ISSUE TITLE
Create semantic coherence scorer for sentences

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#438]
## Goal

Score how semantically coherent a sentence is by measuring average pairwise similarity of its content roots.

## Use Cases

- Filter low-quality corpus sentences
- Detect parsing errors (incoherent parses indicate bad parse)
- Validate generated text
- Quality metric for training data
- Find nonsensical sentences ("La hundo manĝas libron")

## Implementation

Create `scripts/score_semantic_coherence.py`:

```python
#!/usr/bin/env python3
"""Score semantic coherence of Esperanto sentences."""

import torch
from klareco.parser import parse

def compute_coherence(sentence: str, root_embeddings, root_to_idx):
    """
    Compute semantic coherence of sentence.
    
    1. Parse sentence to AST
    2. Extract content roots (exclude function words)
    3. Compute pairwise cosine similarities
    4. Return average similarity as coherence score
    
    High score = roots are semantically related
    Low score = roots are unrelated/nonsensical
    """
    ast = parse(sentence)
    
    # Extract content roots (skip la, kaj, de, etc.)
    roots = extract_content_roots(ast)
    
    if len(roots) < 2:
        return None  # Need at least 2 roots
    
    # Compute pairwise similarities
    similarities = []
    for i in range(len(roots)):
        for j in range(i+1, len(roots)):
            if roots[i] in root_to_idx and roots[j] in root_to_idx:
                emb_i = root_embeddings[root_to_idx[roots[i]]]
                emb_j = root_embeddings[root_to_idx[roots[j]]]
                sim = cosine_similarity(emb_i, emb_j)
                similarities.append(sim)
    
    return sum(similarities) / len(similarities) if similarities else 0.0
```

## Example Usage

```bash
# High coherence (makes sense)
python scripts/score_semantic_coherence.py "La hundo manĝas viandon"
# Score: 0.68 ✓ (hund+manĝ+viand are coherent)

# Low coherence (nonsense)
python scripts/score_semantic_coherence.py "La hundo manĝas libron"
# Score: 0.23 ✗ (hund+manĝ+libr unrelated)

# Medium coherence
python scripts/score_semantic_coherence.py "La kato vidas la hundon"
# Score: 0.41 (kat+vid+hund - moderately related)
```

## Batch Processing

```bash
# Score corpus file
python scripts/score_semantic_coherence.py \
  --corpus data/enhanced_corpus/corpus_with_metadata.jsonl \
  --output data/corpus_coherence_scores.jsonl \
  --threshold 0.2

# Output: sentences with coherence scores
# Filter out low-coherence sentences
```

## Applications

1. **Corpus Cleaning**
   ```python
   # Remove sentences with coherence < 0.15
   # Likely parsing errors or nonsense
   ```

2. **Parse Validation**
   ```python
   # If coherence very low, flag for manual review
   # May indicate parser error
   ```

3. **Training Data Quality**
   ```python
   # Prefer high-coherence sentences for training
   # More likely to be correct, meaningful examples
   ```

4. **Generated Text Validation**
   ```python
   # Score generated translations/paraphrases
   # Reject if coherence too low
   ```

## Features

- [ ] Single sentence scoring
- [ ] Batch corpus processing
- [ ] Histogram of coherence distribution
- [ ] Filter by threshold
- [ ] Handle function words correctly (exclude from scoring)

## Deliverables

- [ ] `scripts/score_semantic_coherence.py`
- [ ] Coherence distribution analysis of corpus
- [ ] Recommended threshold values
- [ ] Example low/high coherence sentences

## Success Criteria

- Clearly distinguishes sensible vs nonsensical sentences
- Correlation with human judgments
- Fast enough for corpus-scale processing
- Useful threshold values identified

## Related

- #427 - Root embeddings
- #421 - Parse quality filtering (could use coherence)
- #434 - Parse disambiguation (uses similar coherence scoring)
